File: python/fms.py
import os
import numpy as np
from scipy.io import wavfile


def fms(audio_filename):
    """
    Creates fixed-size modulation spectrum (FMS).

    Creates fixed-size modulation spectrum (FMS) as described in
    XXX.
        Usage: psi_m, psi_p = fms(audio_filename)

    audio_filename is expected to be a .wav file with


    Parameters
    ----------
    audio_filename : str
        A .wav file with fs = 16, 32, 48, 22.05 or 44.1k.  Duration needs to
        be at least 60 ms. If file has more than one channel, then channel 1 is
        used.

    Returns
    -------
    psi_m : TODO: psi_m type
        Magnitude FMS.
    psi_p : TODO: psi_p type
        Phase FMS.

    Notes
    -----
    PsiM and PsiP are matrices with size Nmel by 8. They contain the
    magnitude and phase FMS, respectively, as given in Eqn. (6).
    Nmel depends on fs:
    fs      Nmel
    ------   ----
    16k       32
    32k       40
    48k       45
    22.05k    35
    44.1k     44

    For all 5 sample rates, the lower 32 mel bands match exactly. The
    bands above 32 cover the range from 8 kHz up to near the Nyquist
    frequency for the given sample rate.

    # TODO: Add python info here
    Written January 12, 2023 by S. Voran at Institute for Telecommunication
    Sciences in Boulder, Colorado, United States: svoran@ntia.gov
    Written and tested using MATLAB Version: 9.11.0.1809720 (R2021b) Update 1
    Note that Matlab uses one-based indexing, while the papers use
    zero-based indexing
    """
    # Check that .wav file has been specified
    _, ext = os.path.splitext(audio_filename)
    if ext != ".wav":
        raise ValueError("audio_filename is expected to name a .wav file.")

    # Read audio samples and sample rate
    fs, x = load_audio(audio_filename)
    # Extract channel 1
    if len(x.shape) > 1:
        # x = x[:, 0]
        x = x[0]
    x = x[None, :]
    # Number of audio samples available
    Na = x.shape[1]
    # Check for sufficient audio duration
    if Na / fs < 0.060:
        raise ValueError(
            "File contains less than 60 ms of audio which is not suitable for FMS"
        )

    Nw, Ns, Nmel, fu = get_constants(fs)
    # Sample rate fs determines
    #   - Nw (samples per window or frame)
    #   - Ns (stride in samples)
    #   - Nmel (number of mel spectrum samples)
    #   - fu (upper limit of analysis in Hz)

    # Calc. number of frames those samples allow
    Nf = np.floor((Na - Nw) / Ns).astype(int) + 1
    # Set number of samples in DFT
    Nt = 2 * Nw

    # Make matrix Theta which implements filter bank that creates mel specrrum
    Theta = make_theta(fs, Nt, Nmel, fu)

    # Make matrix Phi which implements filterbank that creates modulation
    # spectrum
    Phi = makePhi(fs, Ns, Nf)

    # Generate normalized periodic Hamming Window with Nw samples - Eqn. (1)
    hammingWindow = (0.54 - 0.46 * np.cos(2 * np.pi * np.arange(0, Nw) / Nw)) / (
        0.54 * Nw
    )

    # Create matrix of windowed audio sample frames - Eqn. (1)
    xW = np.zeros((Nf, Nw))
    for f in range(Nf):
        sample_ix = np.arange(f * Ns, f * Ns + Nw)
        xW[f, :] = np.multiply(x[0, sample_ix], hammingWindow)

    # Zero pad matrix so frames have length Nt  - Eqn. (2)
    xTildeW = np.concatenate([xW, np.zeros((Nf, Nt - Nw))], axis=1)

    # DFT all frames - Eqn. (3)
    # X = fft(xTildeW, [], 2)
    X = np.fft.fft(xTildeW, axis=1)

    # Convert to mel spectrum - Eqn. (4)
    ss_ix = np.arange(Nt / 2 + 1).astype(int)
    P = np.matmul(np.abs(X[:, ss_ix]), Theta)
    # P = np.abs( X[:, 1:Nt/2 + 1] ) * Theta

    # Generate unnormalized symmetric Hamming Window with Nf samples - Eqn. (5)
    hammingWindow = 0.54 - 0.46 * np.cos(2 * np.pi * np.arange(0, Nf) / (Nf - 1))

    # Window and DFT all mel bins across frames - Eqn. (5)
    Gamma = np.fft.fft(
        np.multiply(P, hammingWindow[:, np.newaxis]), axis=0
    ).T.conjugate()

    # Number of Hz-scale spectral samples produced by DFT
    N = np.floor(Nf / 2) + 1
    # Filter DFT result to create modulation spectrum - Eqn. (6)
    N_ix = np.arange(N).astype(int)
    PsiM = np.matmul(np.abs(Gamma[:, N_ix]), Phi)
    PsiP = np.matmul(np.angle(Gamma[:, N_ix]), Phi)

    # (Straight from Matlab warning): Unwrapped phase might be a good option to
    # try in some cases
    # PsiP = unwrap( angle( Gamma(:,1:N) ) )*Phi
    return PsiM, PsiP


def load_audio(audio_filename):
    """
    Load floating point representation of audio file.

    Parameters
    ----------
    audio_filename : str
        Path to audio file.

    Returns
    -------
    fs : int
        Sample rate.
    x : np.ndarray
        Floating point representation of audio file.

    Raises
    ------
    RuntimeError
        Audio is not unsigned 8 bit int, signed 16 or 32 bit int, or floating
        point.
    """
    fs, x = wavfile.read(audio_filename)
    # Convert x to float representation (between -1 and 1)
    if np.issubdtype(x.dtype, np.floating):
        x = x
    elif x.dtype is np.dtype("uint8"):
        x = (x.astype("float") - 64) / 64
    elif x.dtype is np.dtype("int16"):
        x = x.astype("float") / (2**15)
    elif x.dtype is np.dtype("int32"):
        x = x.astype("float") / (2**31)
    else:
        raise RuntimeError(f"unknown audio type '{x.dtype}'")
    return fs, x


def make_theta(fs, Nt, Nmel, fu):
    """
    Make matrix that implements a filter bank that creates mel spectrum.

    Parameters
    ----------
    fs : int
        Audio sample rate
    Nt : int
        Number of samples produced by DFT (must be even)
    Nmel : int
        Number of mel spectrum samples that will be produced
    fu : float
        Upper frequency limit of the analysis in Hz

    Returns
    -------
    Theta : np.ndarray
        NHertz by Nmel matrix representing filter bank that creates mel
        spectrum. NHertz = Nt/2 + 1.
    """
    # Convert upper analysis limit from Hz to mel, GWEMS filter paper Eqn. (1)
    fTildeU = 2595 * np.log10(1 + fu / 700)

    # %Find mel interval, GWEMS filter paper Eqn. (2)
    deltaTilda = fTildeU / (Nmel + 1)

    # Find band limits and centers in mel
    bTilda = deltaTilda * np.arange(0, Nmel + 2)

    # %Convert to Hz, GWEMS filter paper Eqn. (3)
    # b = 700 * (10.0 ^ (bTilda / 2595) - 1)
    b = 700 * (np.power(10.0, (bTilda / 2595)) - 1)

    # %Calculate DFT frequencies in Hz, GWEMS filter paper Eqn. (5)
    f = np.arange(0, Nt / 2 + 1) * fs / Nt

    # %Calculate filter normalizations, GWEMS filter paper Eqn. (6)
    # eta = 1./( b(3:Nmel+2) - b(1:Nmel) )
    eta = np.divide(1, b[2:] - b[:-2])

    # Number of DFT samples
    NHertz = Nt / 2 + 1

    # Convert to ints to make python happy
    NHertz = int(NHertz)
    Nmel = int(Nmel)
    # Calculate filterbank, GWEMS filter paper Eqn. (4)
    Theta = np.zeros((NHertz, Nmel))

    for i in range(Nmel):
        for k in range(NHertz):
            if b[i] <= f[k] and f[k] < b[i + 1]:
                # Lower slope
                Theta[k, i] = eta[i] * (f[k] - b[i]) / (b[i + 1] - b[i])
            elif b[i + 1] <= f[k] and f[k] < b[i + 2]:
                # Upper slope
                Theta[k, i] = eta[i] * (1 - (f[k] - b[i + 1]) / (b[i + 2] - b[i + 1]))
            else:
                Theta[k, i] = 0

    return Theta


def makePhi(fs, Ns, Nf):
    """
    Make matrix Phi which implements a filterbank that creates modulation
    spectrum

    Parameters
    ----------
    fs : int
        Audio sample rate
    Ns : int
        Stride of framing in samples (frame rate is fs/Ns frames per sec)
    Nf : int
        Number of frames to be processed

    Returns
    -------
    Phi : np.ndarray
        N by Nmod, where N = floor(Nf/2) + 1 and Nmod is 8
    """
    # number of spectral samples available
    N = np.floor(Nf / 2).astype(int) + 1
    # set number of modulation spectrum samples
    Nmod = 8

    # Calculate log-scale frequency interval, GWEMS filter paper Eqn. (8)
    DeltaBar = (np.log2(128) - np.log2(4)) / (Nmod - 1)

    # Calculate log-scale initial filter center frequencies,
    # GWEMS filter paper Eqn. (9)
    bBar = np.log2(4) + np.arange(0, Nmod) * DeltaBar

    # Calculate DFT bin spacing, GWEMS filter paper Eqn. (10)
    Deltah = fs / (Ns * Nf)

    # Ignore warning on log2(0)
    with np.errstate(divide="ignore"):
        # Calculate DFT bin log frequencies, GWEMS filter paper Eqn. (11)
        fBar = np.log2(np.arange(0, N) * Deltah)

    # Adjust log-scale initial filter center frequencies to fall on DFT bins,
    # GWEMS filter paper Eqn. (12)
    # bBarPrime = np.log2( np.round( ( 2.^bBar )/Deltah ) * Deltah );
    bBarPrime = np.log2(np.round(np.power(2, bBar) / Deltah) * Deltah)

    # Calculate filter half-widths, GWEMS filter paper Eqn. (13)
    deltaBar = DeltaBar / (2 - np.sqrt(2))

    # %Calculate filter weights to account for number of DFT samples spanned
    # %by each filter, GWEMS filter paper Eqn. (15)
    nu = []
    for m in range(Nmod):
        cond1 = -deltaBar <= (fBar - bBarPrime[m])
        cond2 = (fBar - bBarPrime[m]) < deltaBar
        sum = np.sum(cond1 & cond2)
        if sum == 0:
            val = np.inf
        else:
            val = 1 / sum

        nu.append(val)

    # Calculate filterbank, GWEMS filter paper Eqn. (14)
    Phi = np.zeros((N, Nmod))

    for m in range(Nmod):
        for k in range(N):
            if (bBarPrime[m] - deltaBar <= fBar[k]) and (fBar[k] < bBarPrime[m]):
                # Lower slope
                Phi[k, m] = nu[m] * (fBar[k] - (bBarPrime[m] - deltaBar)) / deltaBar
            elif (bBarPrime[m] <= fBar[k]) and (fBar[k] < bBarPrime[m] + deltaBar):
                # Upper slope
                Phi[k, m] = nu[m] * (1 - (fBar[k] - bBarPrime[m]) / deltaBar)
            else:
                Phi[k, m] = 0
    return Phi


def hz2mel(hz):
    """
    Convert Hz to mel scale

    Parameters
    ----------
    hz : float
        Frequency in hertz.

    Returns
    -------
    float
        Hertz value on the mel scale.
    """
    return 2595 * np.log10(1 + hz / 700)


def mel2hz(mel):
    """
    Convert mel scale to Hz.

    Parameters
    ----------
    mel : float
        Frequency in mel scale

    Returns
    -------
    float
        Frequency in Hz.
    """
    return 700 * (10 ** (mel / 2595) - 1)


def get_constants(fs):
    """
    Return sample rate dependent FMS constants.

    Parameters
    ----------
    fs : int
        Sample rate of audio sample.

    Returns
    -------
    Nw : int
        Samples per window or frame
    Ns : int
        Stride in samples
    Nmel : int
        Number of mel spectrum samples
    fu : float
        Upper limit of analysis in Hz.

    Notes
    -----
    Audio sample rate is used to look up Nw and Ns.
    Audio sample rate is used calculate Nmel and fu.
    """
    # Select paramters based on sample rate (from Table 2)
    if fs == 16000:
        Nw = 256
        Ns = 48
    elif fs == 32000:
        Nw = 512
        Ns = 96
    elif fs == 48000:
        Nw = 768
        Ns = 144
    elif fs == 22050:
        Nw = 384
        Ns = 66
    elif fs == 44100:
        Nw = 768
        Ns = 132
    else:
        raise ValueError(
            f"Unexepcted sample rate {fs}, (16k, 32, 48, 22.05, and 44.1k are expected)."
        )
    # Number of mel spectrum samples desired in DC to 8 kHz
    nWBmel = 32

    # Convert 8 kHz to mel
    mel8k = hz2mel(8000)

    # Width of mel interval so that nWBmel intervals cover DC to 8 kHz
    melDelta = mel8k / (nWBmel + 1)

    # Convert Nyquist frequency to mel
    melNyquist = hz2mel(fs / 2)

    # How many full melDelta intervals will fit into Nyquist band?
    Nmel = np.floor(melNyquist / melDelta) - 1

    # Find upper analysis limit in mel
    fuMel = melDelta * (Nmel + 1)
    fu = mel2hz(fuMel)

    return Nw, Ns, Nmel, fu

File: matlab/FMS.m
function [PsiM, PsiP] = FMS(audioFilename)
% Creates fixed-size modulation spectrum (FMS) as described in
% Voran, S. and Pieper, J. "A Powerful, Fixed-Size Modulation 
% Spectrum Representation for Perceptually-Consistent Speech Evaluation."
% NTIA Tech Memo 24-57*
% Usage: [PsiM, PsiP] = FMS(audioFilename)
%
% audioFilename is expected to be a .wav file with fs = 16, 24, 32, 48,
% 22.05 or 44.1k.  Three seconds of audio are needed so files with duration
% less than three seconds will be zero padded to 3 seconds. If file has
% more than one channel, then channel 1 is used.
%
% PsiM and PsiP are matrices with size Nmel by 11. They contain the
% magnitude and phase FMS, respectively, as given in Eqn. (2).
% Nmel depends on fs:
%  fs      Nmel
%-------   ----
% 16k       32
% 24k       36
% 32k       40
% 48k       45
% 22.05k    35 
% 44.1k     44
%
% For all 6 sample rates, the lower 32 mel bands match exactly. The 
% bands above 32 cover the range from 8 kHz up to near the Nyquist
% frequency for the given sample rate.
%
% Written April 24, 2024 by S. Voran at Institute for Telecommunication
% Sciences in Boulder, Colorado, United States: svoran@ntia.gov
%
% Written and tested using MATLAB Version: 9.11.0.1809720 (R2021b) Update 1
%
% Note that Matlab uses one-based indexing, while the Tech. Memo uses
% zero-based indexing

%Check that .wav file has been specified
[~, ~, ext] = fileparts(audioFilename);
if ~strcmp(ext,'.wav')
    error('audioFilename is expected to name a .wav file')
end

[x,fs] = audioread(audioFilename); %read audio samples and sample rate
x = x(:,1); %extract channel 1

%Check for sufficient audio duration 

if length(x)/fs < 3
    warning(['Audio file is less than 3 sec. long, signal will ' ...
        'be zero padded to 3.0 sec.'])
    nShort = length(x) - 3*fs;  %Number of zeros needed
    x = [x;zeros(nShort,1)];    %Add needed zeros
end

[Nw, Ns, Nmel, fu] = getConstants(fs);
%Sample rate fs determines
%  - Nw (samples per window or frame)
%  - Ns (stride in samples)
%  - Nmel (number of mel spectrum samples)
%  - fu (upper limit of analysis in Hz)

Na = length(x);             %Find number of audio samples available
Nf = floor( (Na-Nw)/Ns )+1;   %Calc. number of frames those samples allow
Nt = 2*Nw;                  %Set number of samples in DFT

%Make matrix Theta which implements filterbank that creates mel spectrum
Theta = makeTheta(fs, Nt, Nmel, fu); %Appendix A

%Make matrix Phi which implements filterbank that creates modulation
%spectrum
Phi = makePhi(fs, Ns, Nf); %Appendix B

%Generate normalized periodic Hamming Window with Nw samples - Eqn. (1)
hammingWindow = ( 0.54-0.46*cos( 2*pi*[0:Nw-1]/Nw ) )/(0.54*Nw);

%Create matrix of windowed audio sample frames - Eqn. (1)
xW = zeros(Nf,Nw);
for f = 0:Nf-1 %For each frame
    xW(f+1,:) = x(f*Ns + 1:f*Ns + Nw)' .* hammingWindow;
end

%Zero pad matrix so frames have length Nt  - Eqn. (2)
xTildeW = [xW, zeros(Nf,Nt-Nw)];

%DFT all frames - Eqn. (3)
X = fft(xTildeW, [], 2);

%Convert to mel spectrum - Eqn. (4)
P = abs( X(:, 1:Nt/2 + 1) ) * Theta;

%Generate unnormalized symmetric Hamming Window with Nf samples - Eqn. (5)
hammingWindow = ( 0.54  -0.46*cos( 2*pi*[0:Nf-1]/(Nf-1) ) );

%Window and DFT all mel bins across frames - Eqn. (5)
Gamma = fft(P.*hammingWindow',[],1)';

N = floor(Nf/2) + 1; %Number of Hz-scale spectral samples produced by DFT

%Filter DFT result to create modulation spectrum - Eqn. (6)
PsiM = abs(   Gamma(:,1:N) )*Phi;
PsiP = angle( Gamma(:,1:N) )*Phi;

%Unwrapped phase might be a good option to try in some cases
%PsiP = unwrap( angle( Gamma(:,1:N) ) )*Phi;

%--------------------------------------------------------------------------
function [Nw, Ns, Nmel, fu] = getConstants(fs)
%Uses audio sample rate fs to look up
%  - Nw (samples per window or frame)
%  - Ns (stride in samples)
%Uses audio sample rate fs to calculate
%  - Nmel (number of mel spectrum samples)
%  - fu (upper limit of analysis in Hz)

%Select paramters based on sample rate (from Table I)
if fs == 16000
    Nw = 256; %Number of samples per frame
    Ns = 32;  %Stride in samples (advance Ns samples each frame)
elseif fs == 24000
    Nw = 384;
    Ns = 48;
elseif fs == 32000
    Nw = 512;
    Ns = 64;
elseif fs == 48000
    Nw = 768;
    Ns = 96;
elseif fs == 22050
    Nw = 384;
    Ns = 44;
elseif fs == 44100
    Nw = 768;
    Ns = 88;
else
    error(['Unexpected sample rate (16, 24, 32, 48, 22.05, and ' ...
        '44.1k are expected).'])
end

%Number of mel spectrum samples desired in DC to 8 kHz 
nWBmel = 32;

%Convert 8 kHz to mel
mel8k = 2595*log10(1+8000/700); 

%Width of mel interval that so that nWBmel intervals cover DC to 8 kHz
melDelta = mel8k/(nWBmel+1); 

%Convert Nyquist frequency to mel
melNyquist = 2595*log10(1+(fs/2)/700); 

%How many full melDelta intervals will fit into Nyquist band?
Nmel = floor(melNyquist/melDelta)-1;

%Find upper analysis limit in mel
fuMel = melDelta*(Nmel+1); 
fu = 700*( 10^(fuMel/2595)-1 ); %Convert upper limit to Hz

%--------------------------------------------------------------------------
function Theta = makeTheta(fs, Nt, Nmel, fu)
% Make matrix Theta which implements a filterbank that creates mel spectrum
% fs is audio sample rate
% Nt is number of samples produced by DFT (must be even)
% Nmel is the number of mel spectrum samples that will be produced
% fu is the upper frequency limit of the analysis in Hz
%
% Theta is NHertz by Nmel, where NHertz =  Nt/2 + 1;

%Convert upper analysis limit from Hz to mel, Eqn. (14)
fTildeU = 2595*log10(1+fu/700); 

%Find mel interval, Eqn. (15)
deltaTilda = fTildeU/(Nmel+1);  %divid

%Find band limits and centers in mel
bTilda = deltaTilda*[0:Nmel+1];

%Convert to Hz, Eqn. (16)
b = 700*(10.^(bTilda/2595)-1);

%Calculate DFT frequencies in Hz, Eqn. (18)
f = [0:Nt/2]*fs/Nt;

%Calculate filter normalizations, Eqn. (19)
eta = 1./( b(3:Nmel+2) - b(1:Nmel) );

NHertz = Nt/2 + 1; %Number of DFT samples

%Calculate filterbank, Eqn. (17)
Theta = zeros(NHertz,Nmel);
for i = 0:Nmel-1 %Loop over bands
    for k = 0:NHertz-1 %Loop over DFT samples
        if ( b(i+1) <= f(k+1) ) && ( f(k+1) < b(i+2) ) %lower slope
          Theta(k+1,i+1) = eta(i+1)*(f(k+1)-b(i+1))/(b(i+2)-b(i+1));

        elseif ( b(i+2) <= f(k+1) ) && ( f(k+1) < b(i+3) ) %upper slope
          Theta(k+1,i+1) = eta(i+1)*(1 - (f(k+1)-b(i+2))/(b(i+3)-b(i+2)) );

        else %filter takes value zero
          Theta(k+1,i+1) = 0;
        end
    end
end

%--------------------------------------------------------------------------
function Phi = makePhi(fs, Ns, Nf)
% Make matrix Phi which implements a filterbank that creates modulation
% spectrum
% fs is audio sample rate
% Ns is stride of framing in samples (frame rate is fs/Ns frames per sec)
% Nf is the number of frames to be processed
%
% Phi is N by Nmod, where N =  floor(Nf/2) + 1 and Nmod is 11

N = floor(Nf/2) + 1; %number of spectral samples available
Nmod = 11; %set number of modulation spectrum samples

%Calculate log-scale frequency interval, Eqn. (21)
DeltaBar = ( log2(128) - log2(0.25) )/(Nmod - 2);

%Calculate log-scale initial filter center frequencies, 
%Eqn. (22)
bBar = log2(.25) + [0:Nmod-2]*DeltaBar;

%Calculate thresholds, Eqn. (23)
delta(1) = -inf;
delta(2:Nmod-1) = (bBar(1:end-1) + bBar(2:end))/2;
delta(Nmod) = inf;

%Calculate DFT bin spacing, Eqn. (24)
Deltah = fs/(Ns*Nf);

%Calculate DFT bin log frequencies, Eqn. (25)
fBar = log2([0:N-1]*Deltah);

%Calculate filter weights to account for number of DFT samples spanned
%by each filter, Eqn. (28)
nu = zeros(1,Nmod-1);
for m = 1:Nmod-1
    nu(m) = 1/sum( ( delta(m) < fBar ) & ...
        ( fBar  <= delta(m+1) ) );
end

%Calculate filterbank
Phi = zeros(N,Nmod);
Phi(1,1) = 1;         %Eqn. (26)

%Eqn. (27)
for m = 1:Nmod-1
         indexList = find( ( delta(m) < fBar ) & ( fBar  <= delta(m+1) ) );
         Phi(indexList,m+1) = nu(m);        
end